core/templates/hashfuncs.h

```diff
@@ -318,225 +297,168 @@ static _FORCE_INLINE_ uint64_t hash_make_uint64_t(T p_in) {
-		return p_lhs.is_same(p_rhs);
-	}
+struct HashMapHasherDefaultImpl<uint64_t> {
+	static _FORCE_INLINE_ uint32_t hash(const uint64_t p_int) { return hash_one_uint64(p_int); }
 };
 
 template <>
-struct HashMapComparatorDefault<Vector4> {
-	static bool compare(const Vector4 &p_lhs, const Vector4 &p_rhs) {
-		return p_lhs.is_same(p_rhs);
-	}
+struct HashMapHasherDefaultImpl<int64_t> {
+	static _FORCE_INLINE_ uint32_t hash(const int64_t p_int) { return hash_one_uint64(uint64_t(p_int)); }
 };
 
 template <>
-struct HashMapComparatorDefault<Rect2> {
-	static bool compare(const Rect2 &p_lhs, const Rect2 &p_rhs) {
-		return p_lhs.is_same(p_rhs);
-	}
+struct HashMapHasherDefaultImpl<float> {
+	static _FORCE_INLINE_ uint32_t hash(const float p_float) { return hash_murmur3_one_float(p_float); }
 };
 
 template <>
-struct HashMapComparatorDefault<AABB> {
-	static bool compare(const AABB &p_lhs, const AABB &p_rhs) {
-		return p_lhs.is_same(p_rhs);
-	}
+struct HashMapHasherDefaultImpl<double> {
+	static _FORCE_INLINE_ uint32_t hash(const double p_double) { return hash_murmur3_one_double(p_double); }
 };
 
 template <>
-struct HashMapComparatorDefault<Plane> {
-	static bool compare(const Plane &p_lhs, const Plane &p_rhs) {
-		return p_lhs.is_same(p_rhs);
-	}
+struct HashMapHasherDefaultImpl<uint32_t> {
+	static _FORCE_INLINE_ uint32_t hash(const uint32_t p_int) { return hash_fmix32(p_int); }
 };
 
 template <>
-struct HashMapComparatorDefault<Transform2D> {
-	static bool compare(const Transform2D &p_lhs, const Transform2D &p_rhs) {
-		return p_lhs.is_same(p_rhs);
-	}
+struct HashMapHasherDefaultImpl<int32_t> {
+	static _FORCE_INLINE_ uint32_t hash(const int32_t p_int) { return hash_fmix32(uint32_t(p_int)); }
 };
 
 template <>
-struct HashMapComparatorDefault<Basis> {
-	static bool compare(const Basis &p_lhs, const Basis &p_rhs) {
-		return p_lhs.is_same(p_rhs);
-	}
+struct HashMapHasherDefaultImpl<uint16_t> {
+	static _FORCE_INLINE_ uint32_t hash(const uint16_t p_int) { return hash_fmix32(uint32_t(p_int)); }
 };
 
 template <>
-struct HashMapComparatorDefault<Transform3D> {
-	static bool compare(const Transform3D &p_lhs, const Transform3D &p_rhs) {
-		return p_lhs.is_same(p_rhs);
+struct HashMapHasherDefaultImpl<int16_t> {
+	static _FORCE_INLINE_ uint32_t hash(const int16_t p_int) { return hash_fmix32(uint32_t(p_int)); }
+};
+
+template <>
+struct HashMapHasherDefaultImpl<uint8_t> {
+	static _FORCE_INLINE_ uint32_t hash(const uint8_t p_int) { return hash_fmix32(uint32_t(p_int)); }
+};
+
+template <>
+struct HashMapHasherDefaultImpl<int8_t> {
+	static _FORCE_INLINE_ uint32_t hash(const int8_t p_int) { return hash_fmix32(uint32_t(p_int)); }
+};
+
+template <typename F, typename S>
+struct HashMapHasherDefaultImpl<Pair<F, S>> {
+	static _FORCE_INLINE_ uint32_t hash(const Pair<F, S> &p_pair) {
+		uint64_t h1 = HashMapHasherDefault::hash(p_pair.first);
+		uint64_t h2 = HashMapHasherDefault::hash(p_pair.second);
+		return hash_one_uint64((h1 << 32) | h2);
+	}
+};
```

Comment:
This should probably be made into a hash function within the Pair template itself. Failing that, the type should be forward-declared.

Reply:
This is one of those cases where it really depends on perspective of which is the more 'fundamental' header. If it is `hashfuncs`, the functions should be declared on `Pair`. If it's `Pair`, they should be declared on `hashfuncs`. I deem `Pair` to be more fundamental.
That being said, forward-declaring `Pair` is much better than including it!